File: HM_Simulator_FEM_FVM_2D/09_HM_Simulator_FEM_FVM_2D/MATLAB/src/Flow_Model/InputFlowData.m
%% - Input function
function [Flow, Gen, Plotting, State, Storage, Wells] = InputFlowData()
%% - General Parameters
%Final time [sec]
Gen.tf = 200;                                                   %[1,1]    
%Time step [sec]
Gen.tstep = 5;                                                  %[1,1]    
%Tolerance [-]
Gen.tol = 1e-5;                                                 %[1,1]   
%Tolerance of sequential coupling [-]
Gen.tol_all = 1e-3;                                             %[1,1]
%Number of cells in x-direction [-]
Gen.Nx = 41;                                                    %[1,1]  
%Number of cells in y-direction [-]
Gen.Ny = 41;                                                    %[1,1]  
%Reservoir length in x-direction [m]
Gen.Lx = 10;                                                    %[1,1]   
%Reservoir length in y-direction [m]
Gen.Ly = 10;                                                    %[1,1]    
%Reservoir height [m]
Gen.Lz = 1;                                                     %[1,1]  
%Initial pressure in reservoir [Pa]
Gen.Pi = 1e7;

%% - Basic Calculations
%Total number of elements
Gen.N = Gen.Nx*Gen.Ny;                                          %[1,1]
%Element edge lengths
Gen.dx = Gen.Lx/Gen.Nx;                                         %[1,1]
%Element edge lengths
Gen.dy = Gen.Ly/Gen.Ny;                                         %[1,1]
%Element edge lengths
Gen.dz = 1;                                                     %[1,1]

% Cell center coordinates
xc = linspace(Gen.dx/2, Gen.Lx-Gen.dx/2, Gen.Nx);
yc = linspace(Gen.dy/2, Gen.Ly-Gen.dy/2, Gen.Ny);

% Mesh of cell centers
[xc,yc] = meshgrid(xc,yc);                        
Storage.x = reshape(xc',1,Gen.Nx*Gen.Ny);                       %[1,N]
Storage.y = reshape(yc',1,Gen.Nx*Gen.Ny);                       %[1,N]

%% - Flow Model
%x-direction
%Permeability, of left edges [m^2]
kx_left = 1e-12;                                                %[1,1] 
%Permeability, of right zone [m^2]
kx_right = 1e-12;                                               %[1,1] 
%Permeability left edge bands length [m]
Lx_k = 2;                                                       %[1,1] 
%Permeability [m^2] 
Flow.kx0 = kx_right*ones(Gen.Nx*Gen.Ny,1);                      %[1,1]
Flow.kx0(Storage.x < Lx_k) = kx_left;

%y-direction
%Permeability, of left edges [m^2]
ky_left = 1e-12;                                                %[1,1] 
%Permeability, of right zone [m^2]
ky_right = 1e-12;                                               %[1,1] 
%Permeability left edge bands length [m]
Ly_k = 2;                                                       %[1,1] 
%Permeability [m^2] 
Flow.ky0 = ky_right*ones(Gen.Nx*Gen.Ny,1);                      %[1,1]
Flow.ky0(Storage.y < Ly_k) = ky_left;

%"Compressibility" of permeability
Flow.ck = 1e-8;                                                 %[1,1] 
%"Compressibility" of due to volumetric strain
Flow.ckv = 1e-8;                                                %[1,1] 
%Reference pressure [Pa]
Flow.kP0 = 1e5;                                                 %[1,1]  

%Porosity, of left edges [-]
phi_left = 0.2;                                                 %[1,1] 
%Porosity, of right zone [-]
phi_right = 0.2;                                                %[1,1] 
%Porosity left edge bands length [m]
L_phi = 2;                                                      %[1,1] 
%Porosity [-]
Flow.phi0 = phi_right*ones(Gen.Nx*Gen.Ny,1);                    %[1,1]  
Flow.phi0(Storage.x < L_phi) = phi_left;

%"Compressibility" of porosity
Flow.cphi = 1e-9;                                               %[1,1] 
%Reference pressure [Pa]
Flow.phiP0 = 1e5;                                               %[1,1] 


%Fluid compressibility [1/Pa]
Flow.cf = 1e-8;                                                 %[1,1]          
%Fluid viscosity [Pa sec]
Flow.muf = .1;                                                  %[1,1]           
%Reference density [kg/m^3]
Flow.Rho0 = 1000;                                               %[1,1]         
%Reference pressure [Pa]
Flow.RhoP = 1e5;                                                %[1,1]  

%% - Wells
% Constant pressure wells
%Row vector of well pressures [Pa]
Wells.P = [3.5e7];                                              %[1,Nwells]
%Well indexes [m]
Wells.WI = [1000];                                              %[1,Nwells]
% Well locations [m]
Wells.xP = [5];                                                 %[1,Nwells]
Wells.yP = [5];                                                 %[1,Nwells]

    
%Find the cells of these wells
Wells.Loc_P = zeros(size(Wells.xP));
for i = 1:length(Wells.xP)
    [~, Wells.Loc_P(i)] = min((Storage.x - Wells.xP(i)).^2 + (Storage.y - Wells.yP(i)).^2);
end

%Constant rate wells (always keep at least a zero contribution in one cell
%Define constant rate [kg/sec]
Wells.Q = 0;                                                    %[1,Nwells_Q]
Wells.xQ = 5;                                                   %[1,Nwells_Q]
Wells.yQ = 5;                                                   %[1,Nwells_Q]

%Find the cells of these wells
Wells.Loc_Q = zeros(size(Wells.xQ));
for i = 1:length(Wells.xQ)
    [~, Wells.Loc_Q(i)] = min((Storage.x - Wells.xQ(i)).^2 + (Storage.y - Wells.yQ(i)).^2);
end


%% - Initialize state
%Initialize time
State.t = 0;                                                    %[1,1]
%Initialize pressure
State.P = zeros(Gen.Nx*Gen.Ny,1) + Gen.Pi;                      %[N,1]

%Time step counter
State.step = 1;                                                 %[1,1]

%% - Plotting inputs
Plotting.lwidth_1col = 0.75;
Plotting.Position_1col_matrix = [2.2 1.8 6 4.5];
Plotting.fsize_1col = 7;




end

File: HM_Simulator_FEM_FVM_2D/09_HM_Simulator_FEM_FVM_2D/Python/src/src_flow/input_flow_data.py
import numpy as np

from src.src_flow.phiCalc import phiCalc
from src.src_flow.perm import perm

def input_flow_data():

    # ------------------------------------------------------------------
    # General Parameters
    # ------------------------------------------------------------------

    Gen = {}

    Gen["tf"] = 20.0  # Final time [sec]
    Gen["tstep"] = 5.0  # Time step [sec]
    Gen["tol"] = 1e-5  # Tolerance of flow simulator [-]
    Gen["tol_all"] = 1e-4  # Tolerance of coupled system [-]

    Gen["Nx"] = 41  # Number of cells in x direction [-]
    Gen["Ny"] = 41  # Number of cells in y direction [-]

    Gen["Lx"] = 10.0  # Reservoir length x [m]
    Gen["Ly"] = 10.0  # Reservoir length y [m]
    Gen["Lz"] = 1.0  # Reservoir height [m]

    Gen["Pi"] = 1e7 #Initial pressure [Pa]

    # ------------------------------------------------------------------
    # Basic Calculations
    # ------------------------------------------------------------------

    Gen["dx"] = Gen["Lx"] / Gen["Nx"]
    Gen["dy"] = Gen["Ly"] / Gen["Ny"]
    Gen["dz"] = Gen["Lz"]

    # Define grid position, cell centers
    xc = np.linspace(Gen["dx"] / 2, Gen["Lx"] - Gen["dx"] / 2, Gen["Nx"])
    yc = np.linspace(Gen["dy"] / 2, Gen["Ly"] - Gen["dy"] / 2, Gen["Ny"])
    # Mesh of cell centers
    xc_grid, yc_grid = np.meshgrid(xc, yc)

    Storage = {}

    Storage["x"] = xc_grid.T.reshape(-1, order='F')
    Storage["y"] = yc_grid.T.reshape(-1, order='F')

    # ------------------------------------------------------------------
    # Flow Model
    # ------------------------------------------------------------------

    Flow = {}

    # Permeability in x direction [m^2]
    #Permeability on left side
    kx_left = 1e-12
    #Permeability on right side
    kx_right = 1e-12
    #Permeability band size on left side
    Lx_k = 0
    #Reference permeability
    Flow["kx0"] = np.zeros(Gen["Nx"]*Gen["Ny"]) + kx_right  # Permeability [m²]
    Flow["kx0"][Storage["x"] < Lx_k ] = kx_left

    #Permeability in y direction [m^2]
    # Permeability on top side
    ky_top = 1e-12
    # Permeability on bot side
    ky_bot = 1e-12
    # Permeability band size on left side
    Ly_k = 2
    # Reference permeability
    Flow["ky0"] = np.zeros(Gen["Nx"]*Gen["Ny"]) + ky_bot  # Permeability [m²]
    Flow["ky0"][Storage["y"] < Ly_k] = ky_top

    #"Compressibility" of permeability
    Flow["ck"] = 1e-8
    #Dependence of permeability on volumetric strain
    Flow["ckv"] = 1e-8
    #Reference pressure [Pa]
    Flow["kP0"] = 1e5

    #Porosity on the left side
    phi_left = 0.2
    #Porosity on the right side
    phi_right = 0.2
    #Porosity band size on left side
    L_phi = 2
    #Reference porosity
    Flow["phi0"] = np.zeros(Gen["Nx"]*Gen["Ny"]) + phi_right  # Porosity [-]
    Flow["phi0"][Storage["x"] < L_phi] = phi_left
    #"Compressibility" of permeability
    Flow["cphi"] = 1e-9
    #Reference pressure [Pa]
    Flow["phiP0"] = 1e5

    Flow["cf"] = 1e-8  # Fluid compressibility [1/Pa]
    Flow["muf"] = 0.1  # Fluid viscosity [Pa.s]

    Flow["Rho0"] = 1000.0  # Reference density [kg/m³]
    Flow["RhoP"] = 1e5  # Reference pressure [Pa]

    # ------------------------------------------------------------------
    # Wells
    # ------------------------------------------------------------------

    #Constant pressure wells
    Wells = {}
    Wells["P"] = [3.5e7] #Row vector of well pressurese [Pa]
    Wells["WI"] = [1000]    #Well indices [m]
    Wells["xP"] = [5]   #Well locations in x [m]
    Wells["yP"] = [5]  #Well locations in y [m]

    #Find the cells for these wells
    Wells["Loc_P"] = np.zeros(np.size(Wells["P"]), dtype=int)
    for i in range(len(Wells["xP"])):
        Wells["Loc_P"][i] = np.argmin((Storage["x"] - Wells["xP"][i])**2 + (Storage["y"] - Wells["yP"][i])**2)

    # Constant rate wells
    Wells["Q"] = [0]  # Row vector of well rates [kg/sec]
    Wells["xQ"] = [5]  # Well locations in x [m]
    Wells["yQ"] = [5]  # Well locations in y [m]

    # Find the cells for these wells
    Wells["Loc_Q"] = np.zeros(np.size(Wells["Q"]), dtype=int)
    for i in range(len(Wells["xQ"])):
        Wells["Loc_Q"][i] = np.argmin((Storage["x"] - Wells["xQ"][i])**2 + (Storage["y"] - Wells["yQ"][i])**2)


    # ------------------------------------------------------------------
    # Initialize State
    # ------------------------------------------------------------------

    State = {}

    State["t"] = 0.0

    State["P"] = np.full(Gen["Nx"]*Gen["Ny"],Gen["Pi"])

    State["e_vol"] = np.full(Gen["Nx"] * Gen["Ny"], 0)

    State["step"] = 0

    State["kx"],State["ky"],_,_,_,_ = perm(Flow,State)

    State["phi"],_ = phiCalc(Flow,State)



    # =============================================================================
    # Plotting parameters
    # =============================================================================

    Plotting = {}
    Plotting["lwidth_1col"] = 0.75
    Plotting["Position_1col_matrix"] = [2.2, 1.8, 6, 4.5]
    Plotting["fsize_1col"] = 7

    return Flow, Gen, Plotting, State, Storage, Wells